### PR TITLE
Added configurable request timeout to API v5 client

### PR DESCRIPTION
**Pull request by Opelord**
Issue: KNTK-329

### CODE REVIEW

File: kentikapi/internal/api_connection/rest_client.go [hunk 1]
```diff
@@ -124,8 +161,12 @@ func (c *RestClient) Put(ctx context.Context,
 	return body, errorFromResponseStatus(response, string(body))
 }
 
-// Delete sends DELETE request to the API and returns raw response body.
+// Delete sends DELETE request to the API and returns raw response body
```
Comment (mmac-m3a) [inline kentikapi/internal/api_connection/rest_client.go:164]: Shouldn't there really be a period at the end of the sentence? Doesn't one of the linters insist on that?

**Review by mmac-m3a — DISMISSED** [review]
LGTM

**Comment by mmac-m3a** [comment]
I'll wait for @Fenthick to approve too before merging.

File: kentikapi/users_integration_test.go [hunk 2]
```diff
@@ -194,6 +195,108 @@ func TestClient_GetAllUsers(t *testing.T) {
...
+			// assert
+			t.Logf("Got result: %v, err: %v", result, err)
+			if tt.expectedError {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+
+			if tt.expectedTimeout {
+				assert.True(t, strings.Contains(err.Error(), "context deadline exceeded"))
```
Comment (Fenthick) [inline kentikapi/users_integration_test.go:282]: "context deadline exceeded" string is an implementation detail of the _context_ package, so we should avoid relying on that, even in tests.
Please use something like `assert.True(t, errors.Is(err, context.DeadlineExceeded))`.

Using errors.As() ensures that test will not break when our implementation wraps the error.
Using context.DeadlineExceeded ensures that test will not break when error message changes.

See: https://go.dev/blog/go1.13-errors

File: kentikapi/users_integration_test.go [hunk 3]
```diff
@@ -194,6 +195,108 @@ func TestClient_GetAllUsers(t *testing.T) {
...
+				testutil.NewErrorHTTPResponse(http.StatusBadGateway),
+				testutil.NewErrorHTTPResponse(http.StatusBadGateway),
+				{StatusCode: http.StatusBadRequest, Body: `{"error":"Bad Request"}`},
+			},
+			expectedError:   true,
+			expectedTimeout: false,
+			timeout:         durationPtr(10 * time.Second),
+			expectedReqs:    3,
+		}, {
+			name: "timeout after 1 request",
```
Comment (Fenthick) [inline kentikapi/users_integration_test.go:240]:
```suggestion
			name: "timeout during first request",
```
First request won't even finish, so "after" is not precise

File: kentikapi/users_integration_test.go [hunk 4]
```diff
@@ -194,6 +195,108 @@ func TestClient_GetAllUsers(t *testing.T) {
...
+		responses       []testutil.HTTPResponse
+		expectedResult  *models.User
+		expectedError   bool
+		expectedTimeout bool
+		// timeout and expected requests are strictly connected. The server sleeps for 1 ms before each response.
+		timeout      *time.Duration
+		expectedReqs int
+	}{
+		{
+			name: "default timeout",
```
Comment (Fenthick) [inline kentikapi/users_integration_test.go:210]:
```suggestion
			name: "default timeout is larger than 30 ms",
```

File: kentikapi/users_integration_test.go [hunk 5]
```diff
@@ -194,6 +195,108 @@ func TestClient_GetAllUsers(t *testing.T) {
...
+			responses: []testutil.HTTPResponse{
+				testutil.NewErrorHTTPResponse(http.StatusBadGateway),
+				testutil.NewErrorHTTPResponse(http.StatusBadGateway),
+				{StatusCode: http.StatusBadRequest, Body: `{"error":"Bad Request"}`},
+			},
+			expectedError:   true,
+			expectedTimeout: false,
+			expectedReqs:    3,
+		}, {
+			name: "status bad request",
```
Comment (Fenthick) [inline kentikapi/users_integration_test.go:220]: I don't see additional value in this test case when comparing to _TestClient_GetUser.status bad request_. Please remove that test case.

File: kentikapi/users_integration_test.go [hunk 6]
```diff
@@ -194,6 +195,108 @@ func TestClient_GetAllUsers(t *testing.T) {
 	}
 }
 
+func TestClient_GetUser_WithTimeout(t *testing.T) {
```
Comment (Fenthick) [inline kentikapi/users_integration_test.go:198]: It think that this test could be easily merged with TestClient_GetUser.

1. Provide serverHandlingDelay duration as test parameter and pass it to test HTTP handler.
2. Remove expectedReqs and just skip checking number of requests if expectedTimeout is true.

Please merge the tests or explain why they are separate.

**Review by Fenthick — CHANGES_REQUESTED** [review]
Good job overall (nice tests!), but we need to reduce duplication in tests.

File: kentikapi/internal/testutil/http.go [hunk 7]
```diff
@@ -63,12 +59,17 @@ type MultipleResponseSpyHTTPHandler struct {
 
 	// Requests spied by the handler
 	Requests []HTTPRequest
+
+	// handlingDelay delays the response for of SpyHTTPHandler
```
Comment (Fenthick) [inline kentikapi/internal/testutil/http.go:63]:
```suggestion
	// handlingDelay specifies the delay applied while handling the request.
```

File: kentikapi/internal/testutil/http.go [hunk 8]
```diff
@@ -88,6 +89,7 @@ func (h *MultipleResponseSpyHTTPHandler) ServeHTTP(rw http.ResponseWriter, r *ht
 
 	rw.Header().Set("Content-Type", "application/json")
 	response := h.response()
+	time.Sleep(h.handlingDelay)
```
Comment (Fenthick) [inline kentikapi/internal/testutil/http.go:92]: Nit: I'd prefer that line above starting forming the response (i.e. above `rw.Header().Set("Content-Type", "application/json")`), not in the middle of it, to increase the cohesion of the block of code. Possible with a line break between them.

File: kentikapi/users_integration_test.go [hunk 9]
```diff
@@ -195,114 +195,17 @@ func TestClient_GetAllUsers(t *testing.T) {
...
 
-func TestClient_GetUser_WithTimeout(t *testing.T) {
+func TestClient_GetUser(t *testing.T) {
 	tests := []struct {
-		name            string
-		responses       []testutil.HTTPResponse
+		name                string
+		responses           []testutil.HTTPResponse
+		serverHandlingDelay time.Duration
+		// timeout and expected requests are strictly connected. The server sleeps for 10 ms before each response.
```
Comment (Fenthick) [inline kentikapi/users_integration_test.go:203]:
```suggestion
```
Please just remove remove that comment. The second sentence is false for multiple test cases. And the first one doesn't explain anything (at least for me)

**Review by Fenthick — CHANGES_REQUESTED** [review]

File: kentikapi/internal/api_connection/rest_client.go (hunk 1 above)
Comment (Fenthick) [reply kentikapi/internal/api_connection/rest_client.go]: Done

File: apiv6/kentikapi/httputil/httputil_test.go [hunk 10]
```diff
@@ -119,7 +123,3 @@ func (t *spyTransport) RoundTrip(_ *http.Request) (*http.Response, error) {
 func intPtr(v int) *int {
 	return &v
 }
-
-func durationPtr(v time.Duration) *time.Duration {
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply apiv6/kentikapi/httputil/httputil_test.go]: Done

File: kentikapi/internal/api_connection/rest_client.go [hunk 11]
```diff
@@ -16,21 +17,40 @@ import (
 const (
 	authEmailKey    = "X-CH-Auth-Email"
 	authAPITokenKey = "X-CH-Auth-API-Token"
+	defaultTimeout  = 100 * time.Second
 )
 
 type RestClient struct {
 	config     RestClientConfig
 	httpClient *retryablehttp.Client
+	// Timeout specifies a time limit for requests made by this Client. The timeout includes connection time, any
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply kentikapi/internal/api_connection/rest_client.go]: Done

File: kentikapi/internal/api_connection/rest_client.go [hunk 12]
```diff
@@ -16,21 +17,40 @@ import (
...
+	// redirects, and reading the response body. The timer remains running after Get, Head, Post, or Do return and will
+	// interrupt reading of the Response.Body. A Timeout of zero means no timeout.
+	Timeout *time.Duration
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply kentikapi/internal/api_connection/rest_client.go]: Done

File: kentikapi/internal/api_connection/rest_client.go [hunk 13]
```diff
@@ -16,21 +17,40 @@ import (
...
 	authAPITokenKey = "X-CH-Auth-API-Token"
+	defaultTimeout  = 100 * time.Second
 )
 
 type RestClient struct {
 	config     RestClientConfig
 	httpClient *retryablehttp.Client
+	// Timeout specifies a time limit for requests made by this Client. The timeout includes connection time, any
+	// redirects, and reading the response body. The timer remains running after Get, Head, Post, or Do return and will
+	// interrupt reading of the Response.Body. A Timeout of zero means no timeout.
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply kentikapi/internal/api_connection/rest_client.go]: Done

File: kentikapi/internal/api_connection/rest_client.go [hunk 14]
```diff
@@ -16,21 +17,40 @@ import (
...
 type RestClientConfig struct {
 	APIURL    string
 	AuthEmail string
 	AuthToken string
 	RetryCfg  httputil.RetryConfig
+	Timeout   *time.Duration
 }
 
 func NewRestClient(c RestClientConfig) *RestClient {
+	if c.Timeout != nil {
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply kentikapi/internal/api_connection/rest_client.go]: Done

File: kentikapi/internal/api_connection/rest_client.go [hunk 15]
```diff
@@ -16,21 +17,40 @@ import (
...
 	authEmailKey    = "X-CH-Auth-Email"
 	authAPITokenKey = "X-CH-Auth-API-Token"
+	defaultTimeout  = 100 * time.Second
 )
 
 type RestClient struct {
 	config     RestClientConfig
 	httpClient *retryablehttp.Client
+	// Timeout specifies a time limit for requests made by this Client. The timeout includes connection time, any
+	// redirects, and reading the response body. The timer remains running after Get, Head, Post, or Do return and will
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply kentikapi/internal/api_connection/rest_client.go]: Done

File: kentikapi/internal/api_connection/rest_client.go [hunk 16]
```diff
@@ -40,11 +60,20 @@ func NewRestClient(c RestClientConfig) *RestClient {
 					RetryCfg:   c.RetryCfg,
 				},
 			),
+		Timeout: durationPtr(defaultTimeout),
 	}
 }
 
+func durationPtr(duration time.Duration) *time.Duration {
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply kentikapi/internal/api_connection/rest_client.go]: Done

File: kentikapi/internal/testutil/http.go [hunk 17]
```diff
@@ -120,3 +125,55 @@ func NewErrorHTTPResponse(statusCode int) HTTPResponse {
 		Body:       fmt.Sprintf(`{"error":"%v"}`, http.StatusText(statusCode)),
 	}
 }
+
+type MultipleResponseSpyHTTPHandlerSleep struct {
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply kentikapi/internal/testutil/http.go]: Done

File: kentikapi/internal/testutil/http.go [hunk 18]
```diff
@@ -6,10 +6,15 @@ import (
 	"net/http"
 	"net/url"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/assert"
 )
 
+const (
+	serverSleepTime = 10 * time.Millisecond
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply kentikapi/internal/testutil/http.go]: Done

File: kentikapi/users_integration_test.go [hunk 19]
```diff
@@ -194,6 +195,108 @@ func TestClient_GetAllUsers(t *testing.T) {
...
 
+func TestClient_GetUser_WithTimeout(t *testing.T) {
+	tests := []struct {
+		name            string
+		responses       []testutil.HTTPResponse
+		expectedResult  *models.User
+		expectedError   bool
+		expectedTimeout bool
+		// timeout and expected requests are strictly connected. The server sleeps for 1 ms before each response.
+		timeout      *time.Duration
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply kentikapi/users_integration_test.go]: Done

File: kentikapi/users_integration_test.go [hunk 20]
```diff
@@ -194,6 +195,108 @@ func TestClient_GetAllUsers(t *testing.T) {
...
 }
 
+func TestClient_GetUser_WithTimeout(t *testing.T) {
+	tests := []struct {
+		name            string
+		responses       []testutil.HTTPResponse
+		expectedResult  *models.User
+		expectedError   bool
+		expectedTimeout bool
+		// timeout and expected requests are strictly connected. The server sleeps for 1 ms before each response.
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply kentikapi/users_integration_test.go]: Done

File: kentikapi/users_integration_test.go (hunk 2 above)
Comment (Fenthick) [reply kentikapi/users_integration_test.go]: Done

File: kentikapi/users_integration_test.go (hunk 3 above)
Comment (Fenthick) [reply kentikapi/users_integration_test.go]: Done

File: kentikapi/users_integration_test.go (hunk 4 above)
Comment (Fenthick) [reply kentikapi/users_integration_test.go]: Done

File: kentikapi/users_integration_test.go (hunk 5 above)
Comment (Fenthick) [reply kentikapi/users_integration_test.go]: Done

File: kentikapi/users_integration_test.go (hunk 6 above)
Comment (Fenthick) [reply kentikapi/users_integration_test.go]: Please apply suggestion number 2.

File: kentikapi/internal/testutil/http.go (hunk 7 above)
Comment (Fenthick) [reply kentikapi/internal/testutil/http.go]: Done

File: kentikapi/internal/testutil/http.go (hunk 8 above)
Comment (Fenthick) [reply kentikapi/internal/testutil/http.go]: Done

File: kentikapi/users_integration_test.go (hunk 9 above)
Comment (Fenthick) [reply kentikapi/users_integration_test.go]: Done

**Review by Fenthick — CHANGES_REQUESTED** [review]
One left to go: https://github.com/kentik/community_sdk_golang/pull/54#discussion_r716675962

File: kentikapi/users_integration_test.go (hunk 6 above)
Comment (Opelord) [reply kentikapi/users_integration_test.go]: Sorry, I misread. On it.

**Review by Fenthick — APPROVED** [review]

**Review by mmac-m3a — APPROVED** [review]
LGTM

**Pull request merged**